src/atc/utils/SelectAndCastColumns.py

```diff
@@ -0,0 +1,36 @@
+import pyspark.sql.functions as F
+import pyspark.sql.types as T
+from pyspark.sql import DataFrame
+
+
+def SelectAndCastColumns(
+    *, df: DataFrame, schema: T.StructType, caseInsensitiveMatching: bool = False
+) -> DataFrame:
+    """
+    Select and cast columns based pyspark schema.
+    In case columns from shema are missing, they will be added with None values.
+
+    param df: The pyspark DataFrame.
+    param schema: pyspark schema to use for select and cast.
+    param caseInsensitiveMatching: Boolean used for case insensitive matching,
+    by default False.
+
+    return: DataFrame mathing schema
+    """
+
+    if caseInsensitiveMatching:
+        selectAndCastColumnsList = [
+            F.col(c.name).cast(c.dataType).alias(c.name)
+            if c.name.lower() in [col.lower() for col in df.columns]
+            else F.lit(None).cast(c.dataType).alias(c.name)
+            for c in schema
+        ]
+    else:
+        selectAndCastColumnsList = [
+            F.col(c.name).cast(c.dataType).alias(c.name)
+            if c.name in df.columns
+            else F.lit(None).cast(c.dataType).alias(c.name)
+            for c in schema
+        ]
```

Comment:
I think that there is some redundant code here. I could not figure out a good way to avoid the redundant code without making a lot of new code. But here is an outline. You can use it if you feel for it! I dont know if it is good code practice....

```python
def _lower(col, caseInsensitiveMatching):
   return col.lower() if caseInsensitiveMatching else col

selectAndCastColumnsList = [
            F.col(c.name).cast(c.dataType).alias(c.name)
            if _lower(c.name,caseInsensitiveMatching) in [_lower(col,caseInsensitiveMatching) for col in df.columns]
            else F.lit(None).cast(c.dataType).alias(c.name)
            for c in schema
        ]

```